icu4c/source/i18n/erarules.cpp

```diff
@@ -243,7 +243,7 @@ void EraRules::getStartDate(int32_t eraIdx, int32_t (&fields)[3], UErrorCode& st
 }
 
 int32_t EraRules::getStartYear(int32_t eraIdx, UErrorCode& status) const {
-    int year = MAX_INT32;   // bogus value
+    int year = MAX_INT32/2; // bogus value, /2 so don't overflow with adjustments elsewhere
```

Comment:
I'm not sure this one really makes sense.. This is returned only when eraIdx is out of range, and in that case, an error is already set to status.

Reply:
Without that, setting an era beyond Reiwa produces a negative UDate

Reply:
Whereas we want a date in the far future

Reply:
And there is no error status available for e.g. ucal_set(cal, UCAL_ERA, 237);

Reply:
The error might be swallowed... I think someone us using year returned there without checking error.

Reply:
In some cases JapaneseCalendar has an assert that is triggered in debug builds but otherwise ignores the error.

Reply:
GregorianCalendar does below -
```    protected int handleGetExtendedYear() {
        int year;
        if (newerField(EXTENDED_YEAR, YEAR) == EXTENDED_YEAR) {
            year = internalGet(EXTENDED_YEAR, EPOCH_YEAR);
        } else {
            // The year defaults to the epoch start, the era to AD
            int era = internalGet(ERA, AD);
            if (era == BC) {
                year = 1 - internalGet(YEAR, 1); // Convert to extended year
            } else {
                year = internalGet(YEAR, EPOCH_YEAR);
            }
        }
        return year;
    }
```
This means era other than BC is completely ignored.
JapaneseCalendar overrides this method -

```
    protected int handleGetExtendedYear() {
        // EXTENDED_YEAR in JapaneseCalendar is a Gregorian year
        // The default value of EXTENDED_YEAR is 1970 (Showa 45)
        int year;
        if (newerField(EXTENDED_YEAR, YEAR) == EXTENDED_YEAR &&
            newerField(EXTENDED_YEAR, ERA) == EXTENDED_YEAR) {
            year = internalGet(EXTENDED_YEAR, GREGORIAN_EPOCH);
        } else {
            // extended year is a gregorian year, where 1 = 1AD,  0 = 1BC, -1 = 2BC, etc
            year = internalGet(YEAR, 1)                                     // pin to minimum of year 1 (first year)
                    + ERA_RULES.getStartYear(internalGet(ERA, CURRENT_ERA)) // add gregorian starting year
                    - 1;                                                    // Subtract one because year starts at 1
        }
        return year;
    }
```
Setter for field does not really check the value range, `internaGet(ERA, CURRENT_ERA)` returns the value set by setter. Previous implementation just use the value to access array..

Ha, previous implementation looks like a security bug... accessing array with a given index without checking range in cpp code.

Reply:
Well, that *was* triggered in debug builds.

Reply:
I think we probably should set safe guard in handeGetExtendedYear(), which is one of call site. Currently - 
```
int32_t JapaneseCalendar::handleGetExtendedYear()
{
    // EXTENDED_YEAR in JapaneseCalendar is a Gregorian year
    // The default value of EXTENDED_YEAR is 1970 (Showa 45)
    int32_t year;

    if (newerField(UCAL_EXTENDED_YEAR, UCAL_YEAR) == UCAL_EXTENDED_YEAR &&
        newerField(UCAL_EXTENDED_YEAR, UCAL_ERA) == UCAL_EXTENDED_YEAR) {
        year = internalGet(UCAL_EXTENDED_YEAR, kGregorianEpoch);
    } else {
        UErrorCode status = U_ZERO_ERROR;
        int32_t eraStartYear = gJapaneseEraRules->getStartYear(internalGet(UCAL_ERA, gCurrentEra), status);
        U_ASSERT(U_SUCCESS(status));

        // extended year is a gregorian year, where 1 = 1AD,  0 = 1BC, -1 = 2BC, etc
        year = internalGet(UCAL_YEAR, 1)    // pin to minimum of year 1 (first year)
            + eraStartYear                  // add gregorian starting year
            - 1;                            // Subtract one because year starts at 1
    }
    return year;
}
```
We may simply fallback to use the last era when index is larger than last era, and negative value to first era.

How above just changing EraRules::getStartYear to do this? I think it's still OK to set the status as error although.

Reply:
What I care about is the plain C behavior: ucal_set does not have a way to return error for invalid era. And how it behaved before, for whatever reason, was that ucal_set for an era beyond the last valid one set the calendar to a UDate far in the future. I have clients that depend on that behavior, and the test I added checks for that. Do you have a suggested change that does not require altering the test? Otherwise, this is becoming a different change than what was suggested in the JIRA ticket.

Reply:
OK, in a phone meeting about this, we decided to not explicitly make setting an era beyond the last valid era produce a calendar date far in the future. We may decided to pin that to last valid era, as calendar add already does. So that means: Remove the /2 that addressed overflow in that situation; restore the asserts for invalid era etc., and update the test to not explicitly test set for an invalid era.